src/foam/core/ClassLoader.js

```diff
@@ -17,48 +17,92 @@
 
 foam.CLASS({
   package: 'foam.core',
+  name: 'ModelARequireExtension',
+  refines: 'foam.core.Model',
+
+  methods: [
+    function arequire() {
+      var p = Promise.resolve();
+      if (this.extends) {
+        p = p.then(this.__subContext__.arequire(this.extends));
+      }
+      for (var i = 0, a; a = this.axioms_[i]; i++) {
+        if (a.arequire) {
+          p = p.then(a.arequire());
+        }
+      };
+      return p;
+    },
+  ],
+});
+
+foam.CLASS({
+  package: 'foam.core',
+  name: 'RequiresARequireExtension',
+  refines: 'foam.core.Requires',
+
+  methods: [
+    function arequire() {
+      return this.__subContext__.arequire(this.path);
+    },
+  ],
+});
+
+foam.CLASS({
+  package: 'foam.core',
   name: 'ClassLoader',
 
   exports: [
     'arequire',
   ],
 
+  properties: [
+    {
+      name: 'pending',
+      class: 'Object',
+      factory: function() {
+        return {};
+      },
+    },
+  ],
+
   methods: [
     {
       name: 'arequire',
       class: 'foam.core.ContextMethod',
       code: function(X, modelId) {
-        // TODO: This method only adds requires. Make it add extends,
-        // properties, and anything else that might be needed.
+        if (this.__subContext__.isRegistered(modelId)) return Promise.resolve();
+        if (this.pending[modelId]) return this.pending[modelId];
```

Comment:
I would test that this actually handles loops correctly.  I'm not sure it will, because by returning the previous promise you could end up in a situation where arequire('Foo') is waiting on arequire('Bar') which is waiting on the exisiting promise arequire('Foo');

I think you want to return an empty promise if pending is set.